Ambiguous Bean Resolution in ResteasyReactiveCommonRecorder

### Describe the bug

The `ResteasyReactiveCommonRecorder` throws a `jakarta.enterprise.inject.AmbiguousResolutionException` exception when a resource injects two or more CDI-managed subresources that share a common ancestor type, despite the application's injection site being unambiguous.

```java
  private final Map<String, GreeterResource> languages;

  @Inject
  public LanguageResourceImpl(
      final EnglishGreeterResource english, final SpanishGreeterResource spanish) {
    languages = Map.of("en", english, "es", spanish);
  }
```



### Expected behavior

Because the injection site is unambiguous, there should not be a CDI error.

### Actual behavior

The error occurs in the `ResteasyReactiveCommonRecorder` at initialization time.

Stack trace from the attached reproducer.

```
Caused by: jakarta.enterprise.inject.AmbiguousResolutionException: Beans: [CLASS bean [class=org.acme.english.EnglishGreeterResource, id=Rr45rZiX2NZ97mLMEMXhy6l0Iag], CLASS bean [class=org.acme.spanish.SpanishGreeterResource, id=F5G0eJE2vxMnOAaNWldPXC8JwMw]]
        at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceSupplier(ArcContainerImpl.java:325)
        at io.quarkus.arc.runtime.BeanContainerImpl.beanInstanceFactory(BeanContainerImpl.java:33)
        at io.quarkus.resteasy.reactive.common.runtime.ArcBeanFactory.<init>(ArcBeanFactory.java:15)
        at io.quarkus.resteasy.reactive.common.runtime.ResteasyReactiveCommonRecorder.factory(ResteasyReactiveCommonRecorder.java:26)
        at io.quarkus.runner.recorded.ResteasyReactiveProcessor$setupEndpoints1572724422.deploy_0(Unknown Source)
        at io.quarkus.runner.recorded.ResteasyReactiveProcessor$setupEndpoints1572724422.deploy(Unknown Source)
        ... 9 more

```

### How to Reproduce?

The attached reproducer demonstrates a simple example of this error.

When these beans are not subresources (i.e. remove the Jakarta REST annotations on `GreeterResource`), and the injection site is unchanged, then no ambiguous resolution exception occurs in the recorder.

Adding CDI qualifiers to the types `EnglishGreeterResource` and `SpanishGreeterResource` and the injection site does not work around the issue.

Curiously, adding a type parameter to the subresource parent type `GreeterResource` works around the initialization issue (but causes other issues in Quarkus 3.21.0 - see #47106).

[quarkus-subresource-ambigous-resolution-error.zip](https://github.com/user-attachments/files/19553027/quarkus-subresource-ambigous-resolution-error.zip)

### Output of `uname -a` or `ver`

Darwin pixee-mbp-gilday.local 24.3.0 Darwin Kernel Version 24.3.0: Thu Jan  2 20:24:16 PST 2025; root:xnu-11215.81.4~3/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

openjdk version "21.0.6" 2025-01-21 LTS OpenJDK Runtime Environment Temurin-21.0.6+7 (build 21.0.6+7-LTS) OpenJDK 64-Bit Server VM Temurin-21.0.6+7 (build 21.0.6+7-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.21.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.9 (8e8579a9e76f7d015ee5ec7bfcdc97d260186937) Maven home: /Users/jgilday/.m2/wrapper/dists/apache-maven-3.9.9-bin/33b4b2b4/apache-maven-3.9.9 Java version: 21.0.6, vendor: Eclipse Adoptium, runtime: /Library/Java/JavaVirtualMachines/temurin-21.jdk/Contents/Home Default locale: en_US, platform encoding: UTF-8 OS name: "mac os x", version: "15.3.1", arch: "aarch64", family: "mac"

### Additional information

This issue also occurs in 3.19.4 and earlier versions of Quarkus.

--- a/extensions/resteasy-reactive/rest/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/resource/basic/SubResourceAmbiguousInjectTest.java
+++ b/extensions/resteasy-reactive/rest/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/resource/basic/SubResourceAmbiguousInjectTest.java
@@ -1,0 +1,128 @@
+package io.quarkus.resteasy.reactive.server.test.resource.basic;
+
+import static io.restassured.RestAssured.given;
+import static org.hamcrest.CoreMatchers.is;
+
+import java.util.Map;
+import java.util.function.Supplier;
+
+import jakarta.enterprise.context.RequestScoped;
+import jakarta.inject.Inject;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.PathParam;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.resteasy.reactive.server.test.simple.PortProviderUtil;
+import io.quarkus.test.QuarkusUnitTest;
+
+class SubResourceAmbiguousInjectTest {
+    @RegisterExtension
+    static QuarkusUnitTest testExtension = new QuarkusUnitTest()
+            .setArchiveProducer(new Supplier<>() {
+                @Override
+                public JavaArchive get() {
+                    JavaArchive war = ShrinkWrap.create(JavaArchive.class);
+                    war.addClass(PortProviderUtil.class);
+                    war.addClass(EnglishGreeterResource.class);
+                    war.addClass(SpanishGreeterResource.class);
+                    war.addClass(GreeterResource.class);
+                    war.addClass(LanguageResource.class);
+                    war.addClass(LanguageResourceV2.class);
+                    war.addClass(GreeterResourceV2.class);
+                    war.addClass(EnglishGreeterResource.class);
+                    war.addClass(SpanishGreeterResource.class);
+                    return war;
+                }
+            });
+
+    @Test
+    void basicTest() {
+        given().when().get("/languages/en").then().statusCode(200).body(is("hello"));
+        given().when().get("/languages/v2/es").then().statusCode(200).body(is("hola"));
+    }
+
+    @RequestScoped
+    public static class EnglishGreeterResource implements GreeterResource {
+        @Override
+        public String greeting() {
+            return "hello";
+        }
+    }
+
+    @RequestScoped
+    public static class SpanishGreeterResource implements GreeterResource {
+        @Override
+        public String greeting() {
+            return "hola";
+        }
+    }
+
+    public interface GreeterResource {
+        @GET
+        @Produces(MediaType.TEXT_PLAIN)
+        String greeting();
+    }
+
+    @Path("languages")
+    public static class LanguageResource {
+
+        private final Map<String, GreeterResource> languages;
+
+        @Inject
+        public LanguageResource(
+                final EnglishGreeterResource english, final SpanishGreeterResource spanish) {
+            languages = Map.of("en", english, "es", spanish);
+        }
+
+        @Path("{language}")
+        public GreeterResource locateGreeter(@PathParam("language") final String language) {
+            return languages.get(language);
+        }
+    }
+
+    @Path("languages/v2")
+    public static class LanguageResourceV2 {
+
+        private final Map<String, GreeterResourceV2> languages;
+
+        @Inject
+        public LanguageResourceV2(
+                final EnglishGreeterResourceV2 english, final SpanishGreeterResourceV2 spanish) {
+            languages = Map.of("en", english, "es", spanish);
+        }
+
+        @Path("{language}")
+        public GreeterResourceV2 locateGreeter(@PathParam("language") final String language) {
+            return languages.get(language);
+        }
+    }
+
+    public abstract static class GreeterResourceV2 {
+        @GET
+        @Produces(MediaType.TEXT_PLAIN)
+        public abstract String greeting();
+    }
+
+    @RequestScoped
+    public static class EnglishGreeterResourceV2 extends GreeterResourceV2 {
+        @Override
+        public String greeting() {
+            return "hello";
+        }
+    }
+
+    @RequestScoped
+    public static class SpanishGreeterResourceV2 extends GreeterResourceV2 {
+        @Override
+        public String greeting() {
+            return "hola";
+        }
+    }
+}

--- a/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java
+++ b/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java
@@ -297,7 +297,10 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
                 }
                 clazz.setPath(sanitizePath(path));
             }
-            if (factoryCreator != null) {
+            if (factoryCreator != null && !classInfo.isInterface()) {
+                // Most likely an interface for a sub resource. The ResourceLocatorHandler does not use the factory to create new instances, but uses the result of the sub resource locator method instead
+                // Interfaces therefore do not need a factory here
+                // Otherwise, when having multiple implementations of the interface, an Ambiguous Bean Resolution error occurs, since io.quarkus.arc.runtime.BeanContainerImpl.createFactory is run, even if the factory is never invoked
                 clazz.setFactory(factoryCreator.apply(classInfo.name().toString()));
             }
             Map<String, String> classLevelExceptionMappers = this.classLevelExceptionMappers.get(classInfo.name());

